### PR TITLE
feat(desktop): client-side stall detection for chat tool calls

### PR DESCRIPTION
**Pull request by eulicesl**
## Summary

- **Problem:** A running tool call had only two states — `running` / `completed`. When a tool is slow or hangs, the user sees an indefinite spinner with no signal that anything is wrong and no way to bail out. This is a common source of "the chat is stuck / broken" perception.
- **What changed:** Adds client-side stall detection. A `StallDetector` actor watches two clocks per turn (the inter-event gap and each in-flight tool's own duration) and promotes UI state through `.running → .slow → .stalled` at configurable thresholds. The UI surfaces a "still working…" annotation on slow tools and a message-level **Cancel** banner on stalled ones, wired to the existing `AgentBridge.interrupt()`.
- **What did NOT change (scope boundary):** No bridge/protocol or backend changes — the detector reads the timestamped events the bridge already emits. No tool-cancellation timeouts (it only surfaces UI + offers manual Cancel). No telemetry changes.

## Linked Issue / Bounty

- N/A — independent contribution.

## Root Cause (bug fixes only)

- N/A — new feature (UX reliability affordance).

## How It Works

- `StallDetector` is **pure logic**: the current time is passed in as a parameter on every operation (`step(kind:atMs:)`, `tick(atMs:)`), so it has no wall-clock dependency and is fully deterministic under test. Production wraps it in a periodic `tick` task while a turn is active so promotions still fire during silent gaps when no events arrive.
- `StallThresholds` holds the promotion cutoffs (`slowGapMs` / `stalledGapMs`, default 8s / 20s) as a `Sendable`/`Equatable` struct so they can be tuned later with no behavioral churn.
- `ToolCallStatus` is extended from `{ running, completed }` to `{ running, slow, stalled, completed, failed }` with an `isInFlight` helper. The exhaustive `switch`es that consume it are updated across the chat, onboarding, and task-chat surfaces (this is the bulk of the changed-line count — mechanical enum-extension ripple).
- `ChatProvider` feeds every bridge callback to the detector and maps detector transitions onto the tool-call rows; the stalled state drives the Cancel banner.

## Change Type

- [x] Feature

## Scope

- [x] Desktop app (Swift/macOS)

## Security Impact

- New permissions or capabilities introduced? **No**
- Auth or token handling changed? **No**
- Data access scope changed (screen data, OCR, user data)? **No**
- New or changed network calls? **No**
- Plugin or tool execution surface changed? **No** — Cancel routes to the existing `AgentBridge.interrupt()`. The detector observes only event *timing*, never content.

## Testing

- [x] Built locally — `xcrun swift build -c debug --package-path Desktop` → `Build complete!`
- [x] Manual verification: compile + unit tests on macOS. See "Human Verification" for the runtime gap.
- [x] Existing tests pass
- [x] New tests added — `StallDetectorTests` (13 tests: inter-event promotion, per-tool timers tracked independently, reset-on-event, threshold guard) and `ToolCallStatusTests` (3 tests: the extended-enum + `isInFlight` contract).

## Human Verification

- Verified scenarios: clean debug build against current `main`; 16 new tests pass; full test target compiles.
- Edge cases checked (via tests): a tool's own timer promoting independently of the inter-event gap; multiple parallel tools tracked separately; repeated tick at the same timestamp emitting no duplicate transition; thresholds precondition.
- What I did **not** verify (and why): I did not drive the running app to screenshot each `.slow` / `.stalled` state. Deterministically forcing a real stall end-to-end needs a fake-bridge harness (out of scope for this PR), so runtime UI evidence is a known gap.

## Evidence

- [x] Test output:
  ```
  Test Suite 'StallDetectorTests' passed  — Executed 13 tests, 0 failures
  Test Suite 'ToolCallStatusTests' passed — Executed 3 tests, 0 failures
  Build complete!
  ```
- No runtime screenshot — honest gap (see Human Verification). Happy to add one if a reviewer wants it.

## Docs

- [x] N/A — no docs impact

## Performance, Privacy, and Reliability

- Performance: one actor + a 500ms tick task active only during an in-flight turn; negligible. Privacy: the detector sees event *kind* and *timing* only — never prompt, tool input, or output text. Reliability: strictly additive — when nothing stalls, behavior is identical to today.

## Migration / Backward Compatibility

- Backward compatible? **Yes**
- Migration needed? **No** — the `ToolCallStatus` extension is additive; persistence treats the new in-flight states as transient and never relies on them being present in stored history.

## Risks and Mitigations

- **Risk:** Extending `ToolCallStatus` could miss an exhaustive `switch`. **Mitigation:** Swift's exhaustiveness checking forces every consumer to handle the new cases; the compile is green, and `ToolCallStatusTests` locks the `isInFlight` contract.
- **Risk:** A periodic tick task leaking past turn end. **Mitigation:** the tick task is cancelled via `defer` on turn-scope exit (success or throw).
- **Known limitation:** thresholds (8s / 20s) are first-pass defaults, intended to be tuned against real timing data later.

## AI Disclosure

- Tools used: Claude Code (Claude Opus).
- AI-assisted portions: implementation and this PR description.
- How I verified correctness: ran the local debug build and the unit-test suite (all green); manually reviewed every diff and confirmed the detector never touches message content.

### CODE REVIEW

File: desktop/Desktop/Sources/Providers/ChatProvider.swift [hunk 1]
```diff
@@ -3216,6 +3296,45 @@ BROWSER TABS: when you use the browser (Playwright), on your FIRST browser actio
...
+    ) {
+        guard !transitions.isEmpty,
+              let index = messages.firstIndex(where: { $0.id == messageId })
+        else { return }
+
+        for transition in transitions {
+            guard case .tool(let id, _, let to) = transition else { continue }
+            for i in messages[index].contentBlocks.indices {
+                if case .toolCall(let blockId, let name, let oldStatus, let tuid, let input, let output) = messages[index].contentBlocks[i],
+                   tuid == id, oldStatus.isInFlight {
```
[the remark this answers is not in the document]
Comment (eulicesl) [reply desktop/Desktop/Sources/Providers/ChatProvider.swift]: Confirmed and fixed in 147b9b73c — good find.

Root cause exactly as described: the detector registered `toolUseId`-less tools under `untracked-<name>` while the content block kept `toolUseId: nil`, so `tuid == id` never matched and the transitions were dropped.

Rather than special-case the match, I routed both the registration site and the match in `applyStallTransitions` through one `ChatProvider.stallTrackingId(toolUseId:name:)` helper, so the key is derived identically at both ends and can't diverge again. Added tests locking the derivation (real id passes through; nil falls back to `untracked-<name>`).

**Review by kodjima33 — APPROVED** [review]
Desktop feature: client-side stall detection, well-tested actor pattern. Approve only — feature >500 lines and mergeStateStatus BLOCKED, author needs to rebase.

**Comment by eulicesl** [comment]
Rebased onto the latest `main`, checks green, and all review threads resolved — this one's ready to merge whenever convenient. Thanks for the review!

**Comment by Git-on-my-level** [comment]
Nice work on this — the design is clean and the migration is thorough. A few notes from review:

**Strengths**
- `StallDetector` as a pure-logic actor with time passed as a parameter is exactly the right call for testability. The test suite drives it deterministically and covers the edge cases (idempotency, parallel tools, promotion jumps, completion resets).
- The `isInFlight` migration is comprehensive — every `status == .running` / `.running` pattern match across `ChatProvider`, `TaskChatState`, `TaskChatCoordinator`, `OnboardingChatView`, and `ChatPage` is updated. Easy to miss spots here; you caught them all.
- `stallTrackingId` routing both registration and the `applyStallTransitions` match through one helper is a good guard against silent drops for `toolUseId`-less tools.
- Persistence mapping is sound — collapsing `.slow`/`.stalled` to `"running"` on write (transient states) and decoding all three codes on read, with unknown codes falling back to `.completed` for forward-compat.

**One consideration before ship**
The inter-event gap and the per-tool duration currently share the same thresholds (8s/20s). A tool that legitimately runs for >20s with no intermediate events (e.g. a large `execute_sql`) would trip the stalled banner even though it is healthy. The V1 note acknowledges these need tuning against real distributions; worth confirming the per-tool timer does not false-positive on known-slow tools during that tuning pass.

Minor: the `TaskChatState.addToolActivity` duplicate of `ChatProvider.addToolActivity` is already flagged with a TODO — good to keep on the radar for a DRY pass.

Overall this is a well-scoped reliability affordance that addresses a real "the chat is stuck" perception. Positive signal from automated review.

**Review by Git-on-my-level — DISMISSED** [review]
Approving — clean ship.

The pure-logic actor design with time injected as a parameter is the right call for testability, and the `isInFlight` migration is thorough across every consuming surface (ChatProvider, TaskChatState, TaskChatCoordinator, OnboardingChatView, ChatPage). The `stallTrackingId` helper introduced in the latest commit resolves the nil-`toolUseId` mismatch cleanly — routing both registration and the transition match through one helper is a good guard against silent drops. Persistence mapping is sound (transient states collapse to "running" on write, unknown codes fall back to `.completed` for forward-compat). The `defer`-cancelled tick task is correctly scoped.

Two non-blocking notes for the radar:

1. `.failed` is defined, rendered, and decoded but not yet wired to any production path — worth a follow-up so it doesn't drift from its "terminal failure" intent, or an explicit note that it's reserved for a future change.
2. The shared 8s/20s thresholds for both the inter-event gap and per-tool duration may false-positive on legitimately slow tools (e.g. a large `execute_sql`) during the tuning pass. Separate per-tool thresholds could be worth considering once you have real timing distributions.

Neither blocks ship. Thanks for the thorough PR description and honest testing-gap disclosure.

**Comment by Git-on-my-level** [comment]
Really nice piece of work — the pure-logic `StallDetector` actor with time passed as a parameter is exactly the right design for testability, and the 16 deterministic tests cover the edges (idempotency, parallel tools, promotion jumps, completion resets). The `isInFlight` migration is thorough across every consuming surface (`ChatProvider`, `TaskChatState`, `TaskChatCoordinator`, `OnboardingChatView`, `ChatPage`) — I grepped for leftover `status == .running` / `.running` tool-call patterns and found none missed. Routing both registration and the `applyStallTransitions` match through `stallTrackingId` is a good guard against silent drops for `toolUseId`-less tools, and the persistence round-trip is sound (transient `.slow`/`.stalled` collapse to `"running"` on write; decode falls back to `.completed` for forward-compat).

**One thing to address before merge** (this also came up in cubic's latest review and looks correct to me):

The `catch` block now calls `completeRemainingToolCalls(messageId:..., terminalStatus: .failed)` unconditionally. Since `stopAgent()` → `agentBridge.interrupt()` → `messageContinuation.resume(throwing: BridgeError.stopped)`, a user-initiated Cancel also lands in that catch. That means tapping the new **Cancel** banner marks every in-flight tool row as a red `.failed` icon — even though the stop was intentional, not an execution error. The code already distinguishes `.stopped` from real errors a few lines further down (the "User stopped — no error to show" branch), so gating `terminalStatus` on `.stopped` resolves it cleanly:

```swift
let terminalStatus: ToolCallStatus = {
    if let bridgeError = error as? BridgeError, case .stopped = bridgeError {
        return .completed
    }
    return .failed
}()
completeRemainingToolCalls(messageId: aiMessageId, terminalStatus: terminalStatus)
```

**Non-blocking notes for the radar**
1. `.failed` is defined, rendered, and decoded but not yet wired to any production path other than this catch block — worth a follow-up so it lands in its "terminal failure" intent, or an explicit note that it's reserved.
2. The shared 8s/20s thresholds apply to both the inter-event gap and per-tool duration. A tool that legitimately runs >20s with no intermediate events (e.g. a large `execute_sql`) would trip the stalled banner while healthy — the V1 note acknowledges tuning is needed; worth confirming per-tool timers don't false-positive on known-slow tools during that pass.
3. The `TaskChatState.addToolActivity` duplicate of `ChatProvider.addToolActivity` is already flagged with a TODO — fine for a later DRY pass.

Flagging for maintainer review given the UX-semantic decision above. Overall a well-scoped reliability affordance addressing a real "the chat is stuck" perception — credit to @eulicesl for the thorough PR description and honest testing-gap disclosure.

**Review by Git-on-my-level — COMMENTED** [review]
Thanks @eulicesl — the pure-logic `StallDetector` actor design and the `isInFlight` migration across every consuming surface are really nicely done. Most of the earlier review items (stale grouped-blocks cache, duplicate `toolStarted` reset, wall-clock vs monotonic time, no-op Cancel button guard, `toolStartTimes` keying) look resolved in the latest commit.

**One thing still open in `87cec5a`** (agreeing with cubic's latest run and the note above):

The catch block at `ChatProvider.swift:3003` calls `completeRemainingToolCalls(messageId: aiMessageId, terminalStatus: .failed)` unconditionally. Since `stopAgent()` → `agentBridge.interrupt()` resumes the continuation with `BridgeError.stopped`, a user-initiated tap on the new **Cancel** banner lands in that same catch and marks every in-flight tool row as a red `.failed` icon — the opposite of the clean cancel this feature is adding. Gating `terminalStatus` on `BridgeError.stopped` (collapse to `.completed` / non-failed for intentional stops) before that call would make the banner do what it advertises.

**Non-blocking, for the radar:** the inter-event gap and per-tool duration share the same 8s/20s thresholds today. A tool that legitimately runs long (e.g. a large `execute_sql`) with no intermediate events would trip the stalled banner even when healthy — worth confirming against real distributions during the V1 tuning pass.

Leaving for human maintainer review before merge.
